nntrainer/src/conv2d_layer.cpp

```diff
@@ -138,6 +176,90 @@ sharedConstTensor Conv2DLayer::backwarding(sharedConstTensor derivative,
+             input_dim.width() + padding[1] * 2);
+  ret.setZero();
+
+#ifdef USE_BLAS
+  int status = ML_ERROR_NONE;
+  for (unsigned int b = 0; b < input_dim.batch(); ++b) {
+    std::vector<float> out(kernel_size[0] * kernel_size[1] *
+                           input_dim.channel() * filter_size);
+
+    Tensor inSub(
+      TensorDim(1, input_dim.channel(), input_dim.height(), input_dim.width()),
+      input.getAddress(b * input.getDim().getFeatureLen()));
+
+    status = conv2d_gemm(
+      derivative->getAddress(b * derivative->getDim().getFeatureLen()),
+      TensorDim(1, derivative->channel(), derivative->height(),
+                derivative->width()),
+      inSub,
+      TensorDim(1, 1, filter_size,
+                kernel_size[0] * kernel_size[1] * input_dim.channel()),
+      stride, padding, out.data(), out.size(), false);
+    if (status != ML_ERROR_NONE)
+      throw std::runtime_error("Backwarding Convolution failed.");
+
+    for (unsigned int i = 0; i < filter_size; ++i) {
+      Tensor &delK = paramsAt(i).grad;
+      Tensor &delBias = paramsAt(i + filter_size).grad;
+      float *del = delK.getData();
+      unsigned int s = kernel_size[0] * kernel_size[1] * input_dim.channel();
+
+      for (unsigned int k = 0; k < s; ++k) {
+        del[k] += out[i * s + k];
+      }
+
+      float sum = 0.0f;
+      for (unsigned int j = 0; j < derivative->height(); ++j) {
+        for (unsigned int k = 0; k < derivative->width(); ++k) {
+          sum += derivative->getValue(b, i, j, k);
+        }
+      }
+      delBias.setValue(0, 0, 0, 0, sum + delBias.getValue(0, 0, 0, 0));
+    }
+  }
+
+  TensorDim kdim(ret.channel(), filter_size, kernel_size[0], kernel_size[1]);
+
+  std::vector<float> imKernel(kdim.getDataLen());
+
+  unsigned int count = 0;
+  float *d = imKernel.data();
+
+  for (unsigned int j = 0; j < ret.channel(); ++j) {
+    for (unsigned int i = 0; i < filter_size; ++i) {
+      Tensor &filters = paramsAt(i).weight;
+      for (unsigned int k = 0; k < kernel_size[0] * kernel_size[1]; ++k) {
+        d[count++] = filters.getData()[j * kernel_size[0] * kernel_size[1] + k];
+      }
+    }
+  }
+
+  TensorDim input_dim_padded(1, input_dim.channel(),
+                             input_dim.height() + padding[0] * 2,
+                             input_dim.width() + padding[1] * 2);
+
+  for (unsigned int b = 0; b < input_dim.batch(); ++b) {
+    std::vector<float> out(input_dim_padded.getFeatureLen());
+    Tensor inSub(
+      TensorDim(1, derivative->channel(), derivative->height(),
+                derivative->width()),
+      derivative->getAddress(b * derivative->getDim().getFeatureLen()));
+
+    status = conv2d_gemm(imKernel.data(), kdim, inSub, input_dim_padded, stride,
+                         same_pad, out.data(), out.size(), true);
+    if (status != ML_ERROR_NONE)
+      throw std::runtime_error("Forwarding Convolution failed.");
+    memcpy(ret.getAddress(b * ret.getDim().getFeatureLen()), out.data(),
```

Comment:
how about putting  `ret.getAddress(b * ret.getDim().getFeatureLen())` at line 252 directly in conv2d_gemm.
I see no reason to get the result in out and then make another copy.

Reply:
fixed. Thanks.